### DISCUSSION TITLE
Change in weight-producer to apply weight on DATA

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#555] [General]
Hello Experts,
(@pkausw @mafrahm et.al.)

I have noticed that there is a dedicated `weight-producer` in the latest version of `columnflow`. Regarding that, I have a couple of queries:

 - What is the necessity of this `weight_producer`? If I want to use different weights for different `process_id`, is it possible to do it in this task?
 - I saw that this `producer` is for `mc_only`. In order to apply some specific weights on Data, if I just make `mc_only=False`, and in my `production.main`, if I do not add the `mc` condition, will it work? e.g. I wanted to apply `fake_factor weight` for the sideband events only, for both data and mc, so that in `hist-hook` if I just subtract the mc from data in sideband, I will be able to get the fake-contribution in my signal region.
 - [digressing a bit] I have a set of weights with 5 different variations, and I want to apply all the variations on a single process. Then I have to plot all of those (normalized) on a single plot. Is it possible to achieve this by tuning this `weight_producer` task? Or in addition to `up` and `down`, can I add more systematic variations in config? 

Please let me know if you have any ideas or comments. 
Thank you very much.

Best,
Gourab

**Comment by mafrahm** [comment]
Hi @gsaha009, the `weight_producer` is basically the same as any producer, but it also returns some weight column that is used as weight when filling histograms. Regarding your questions:

- the `all_weights` Producer [1] already allows setting event weights per dataset_inst. If your usecase is not covered from that, you can simply implement your own `weight_producer`
- when setting `mc_only=False`, the `weight_producer` will also be called for data events. I am not too familiar with the `fake_factor` method, but technically that should be possible:
```
weight = ak.where(sideband_mask, sideband_weight, default_weight)
```
- This sounds like you have one nominal weight and two systematic uncertainties with up+down variations. In that case, you could either implement a single combined systematic uncertainty (e.g. by building the envelope of both uncertainties) or you could implement two separate systematic uncertainties. We implemented both of those ways e.g. for the scale uncertainties [2]. Which type of variation are you referring to?


[1] https://github.com/columnflow/columnflow/blob/273b0ba7049e4999b742910ab53a67ad94526233/columnflow/weight/all_weights.py
[2] https://github.com/columnflow/columnflow/blob/273b0ba7049e4999b742910ab53a67ad94526233/columnflow/production/cms/scale.py

**Reply by gsaha009** [reply]
Hi @mafrahm 

Thanks for the explanation.
 - clear
 - Got it. I will try to do something like that
 - I have a follow up question on this. Let me give you the exact problem I have. I have a few signal samples, ggf, zh and wh. Each of these has 7 different `tau_spinner_weights`. Now, if I apply all of these weights on a single signal process, I will have 7 variations out of one single process. In `shift` through `config`, I used to mention two of the variations as up and down, but now I want to use all of the weights. I don't think I can achieve that using `shift`. So, I was thinking if in the `weight_producer` something can be done in order to apply all the weights separately on a single signal process.

Gourab
